### PR TITLE
Avoid more allocations when compiling html5ever

### PR DESCRIPTION
**Pull request by nnethercote**
These three commits reduce the number of allocations performed when compiling html5ever from 13.2M to 10.8M, which speeds up compilation by about 2%.

r? @nrc

### CODE REVIEW

File: src/libsyntax/ext/tt/macro_parser.rs [hunk 1]
```diff
@@ -161,7 +162,7 @@ pub fn count_names(ms: &[TokenTree]) -> usize {
     })
 }
 
-pub fn initial_matcher_pos(ms: Rc<Vec<TokenTree>>, sep: Option<Token>, lo: BytePos)
+pub fn initial_matcher_pos(ms: Vec<TokenTree>, sep: Option<Token>, lo: BytePos)
```
Comment (Mark-Simulacrum) [inline src/libsyntax/ext/tt/macro_parser.rs]: I believe this means that we are now copying/moving the Vec; perhaps we only need `&[TokenTree]` here?

**Comment by leonardo-m** [comment]
Are those changes speeding up the compilation of other programs too? Optimizations should be added only after there's a more general proof of their quality.

File: src/libsyntax/ext/tt/macro_parser.rs (hunk 1 above)
Comment (oli-obk) [inline src/libsyntax/ext/tt/macro_parser.rs]: Cloning a Vec copies all the allocated elements. But it can probably use `memcpy`, so that should be fast.

**Comment by bluss** [comment]
It looks like SmallVec should impl Deref/DerefMut so that the .as_slice()/mut calls become redundant, and it's more of a drop in replacement for Vec.

File: src/libsyntax/ext/tt/macro_parser.rs [hunk 2]
```diff
@@ -284,10 +285,8 @@ pub fn parse(sess: &ParseSess,
              mut rdr: TtReader,
              ms: &[TokenTree])
              -> NamedParseResult {
-    let mut cur_eis = Vec::new();
-    cur_eis.push(initial_matcher_pos(Rc::new(ms.iter()
-                                                .cloned()
-                                                .collect()),
+    let mut cur_eis = SmallVector::zero();
+    cur_eis.push(initial_matcher_pos(ms.iter().cloned().collect(),
```
Comment (oli-obk) [inline src/libsyntax/ext/tt/macro_parser.rs]: this should be a simple `ms.to_owned()`

**Comment by nrc** [comment]
r=me with the nits in macro_parser.rs addressed

**Comment by nnethercote** [comment]
Updated to address comments, including adding `Deref` and `DerefMut` support to `SmallVector`, which made things nicer :)

**Pull request merged**